crates/ruff_python_formatter/src/comments/mod.rs

```diff
@@ -1,2 +1,274 @@
+//! ```javascript
+//! /**
+//!  * Documentation comment
+//!  */
+//! async /* comment */ function Test () // line comment
+//! {/*inline*/}
+//! ```
+//!
+//! This flexibility makes formatting comments challenging because:
+//! * The formatter must consistently place comments so that re-formatting the output yields the same result,
+//!   and does not create invalid syntax (line comments).
+//! * It is essential that formatters place comments close to the syntax the programmer intended to document.
+//!   However, the lack of rules regarding where comments are allowed and what syntax they document requires
+//!   the use of heuristics to infer the documented syntax.
+//!
+//! This module tries to strike a balance between placing comments as closely as possible to their source location
+//! and reducing the complexity of formatting comments. It does so by associating comments per node rather than a token.
+//! This greatly reduces the combinations of possible comment positions, but turns out to be, in practice,
+//! sufficiently precise to keep comments close to their source location.
+//!
+//! Luckily, Python doesn't support inline comments, which simplifying the problem significantly.
+//!
+//! ## Node comments
+//!
+//! Comments are associated per node but get further distinguished on their location related to that node:
+//!
+//! ### Leading Comments
+//!
+//! A comment at the start of a node
+//!
+//! ```python
+//! # Leading comment of the statement
+//! print("test");
+//!
+//! [   # Leading comment of a
+//!     a
+//! ];
+//! ```
+//!
+//! ### Dangling Comments
+//!
+//! A comment that is neither at the start nor the end of a node.
+//!
+//! ```python
+//! [
+//!     # I'm between two brackets. There are no nodes
+//! ];
+//! ```
+//!
+//! ### Trailing Comments
+//!
+//! A comment at the end of a node.
+//!
+//! ```python
+//! [
+//!     a, # trailing comment of a
+//!     b, c
+//! ];
+//! ```
+//!
+//! ## Limitations
+//! Limiting the placement of comments to leading, dangling, or trailing node comments reduces complexity inside the formatter but means,
+//! that the formatter's possibility of where comments can be formatted depends on the AST structure.
+//!
+//! For example, *`RustPython`* doesn't create a node for the `/` operator separating positional only arguments from the other arguments.
+//!
+//! ```python
+//! def test(
+//!     a,
+//!     /, # The following arguments are positional or named arguments
+//!     b
+//! ):
+//!     pass
+//! ```
+//!
+//! Because *`RustPython`* doesn't create a Node for the `/` argument, it is impossible to associate any
+//! comments with it. Meaning, the default behaviour is to associate the `# The following ...` comment
+//! with the `b` argument, which is incorrect. This limitation can be worked around by implementing
+//! a custom rule to associate comments for `/` as *dangling comments* of the `Arguments` node and then
+//! implement custom formatting inside of the arguments formatter.
```

Comment:
In rustfmt, there's a bunch of hacks for these kinds of comments. (I haven't looked at the details).

https://github.com/rust-lang/rustfmt/blob/master/Contributing.md#a-quick-tour-of-rustfmt

https://github.com/rust-lang/rustfmt/blob/master/src/missed_spans.rs

> Our primary tool here is to look between spans for text we've missed. For example, in a function call foo(a, b), we have spans for a and b, in this case, there is only a comma and a single space between the end of a and the start of b, so there is nothing much to do. But if we look at foo(a /* a comment */, b), then between a and b we find the comment.

> At a higher level, Rustfmt has machinery so that we account for text between 'top level' items. Then we can reproduce that text pretty much verbatim. We only count spans we actually reformat, so if we can't format a span it is not missed completely but is reproduced in the output without being formatted. This is mostly handled in [src/missed_spans.rs](https://github.com/rust-lang/rustfmt/blob/master/src/missed_spans.rs). See also FmtVisitor::last_pos in [src/visitor.rs](https://github.com/rust-lang/rustfmt/blob/master/src/visitor.rs).